tools/dart-code-gen/lib/src/field_validator_factory.dart

```diff
@@ -47,6 +47,66 @@ class FieldValidatorFactory {
     /// May return `null` to signify that no validation is required for the given field.
     ///
     factory FieldValidatorFactory.forField(FieldDescriptorProto field, ValidatorFactory factory) {
+        var scalarFactory = ScalarFieldValidatorFactory._forType(field, factory);
+        var repeated = field.label == FieldDescriptorProto_Label.LABEL_REPEATED;
+        if (repeated) {
+            return RepeatedFieldValidatorFactory(factory, field, scalarFactory);
+        } else {
+            return scalarFactory;
+        }
+    }
+
+    /// Generates validator code for the specified field.
+    ///
+    /// The validator obtains the field value via the given [fieldValue] expression.
+    ///
+    /// If any constrains violations are discovered, they are added to
+    /// the [ValidatorFactory.violationList] of the [validatorFactory].
+    ///
+    Code createFieldValidator(Expression fieldValue) => null;
+
+    /// Checks if the validated field is required.
+    ///
+    /// Returns `true` if the field is required and `false` if it is optional.
+    ///
+    bool isRequired() {
+        var options = field.options;
+        return options.hasExtension(Options.required)
+            && options.getExtension(Options.required);
+    }
+
+    /// Creates a new validation rule with the given parameters.
+    Rule newRule(LazyCondition condition, LazyViolation violation) {
+        return Rule._(condition, violation, validatorFactory.violationList);
+    }
+
+    /// Creates a validation for the `(required)` constraint.
+    ///
+    /// The [condition] determines whether or not the field value if set. The conditional expression
+    /// should evaluate into `true` if the field is **NOT** set.
+    ///
+    Rule createRequiredRule(LazyCondition condition) {
+        return newRule(condition, (v) => _requiredMissing());
+    }
+
+    /// Generates an expression which constructs a `ConstraintViolation` for a missing required
+    /// field.
+    Expression _requiredMissing() {
+        return violationRef.call([literalString('Field must be set.'),
+                                  literalString(validatorFactory.fullTypeName),
+                                  literalList([field.name])]);
+    }
+}
+
+/// A [FieldValidatorFactory] for non-`repeated` fields.
+///
+class ScalarFieldValidatorFactory extends FieldValidatorFactory {
```

Comment:
I think the name is a bit misleading as this factory validates both [scalar](https://developers.google.com/protocol-buffers/docs/proto#scalar) and message types.